server/src/main/java/org/elasticsearch/common/util/concurrent/ListenableFuture.java

```diff
@@ -43,100 +43,72 @@
-            notifyListenerDirectly(listener);
-        } else {
-            final boolean run;
-            // check done under lock since it could have been modified and protect modifications
-            // to the list under lock
-            synchronized (this) {
-                if (done) {
-                    run = true;
-                } else {
-                    final ActionListener<V> wrappedListener;
-                    if (threadContext == null) {
-                        wrappedListener = listener;
-                    } else {
-                        wrappedListener = ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
-                    }
-                    if (listeners == null) {
-                        listeners = new ArrayList<>();
-                    }
-                    listeners.add(new Tuple<>(wrappedListener, executor));
-                    run = false;
-                }
-            }
-
-            if (run) {
-                // run the callback directly, we don't hold the lock and don't need to fork!
-                notifyListenerDirectly(listener);
-            }
+            notifyListener(listener);
         }
     }
 
     @Override
     protected void done(boolean ignored) {
-        final List<Tuple<ActionListener<V>, ExecutorService>> existingListeners;
-        synchronized (this) {
-            done = true;
-            existingListeners = listeners;
-            if (existingListeners == null) {
-                return;
-            }
-            listeners = null;
-        }
-        for (Tuple<ActionListener<V>, ExecutorService> t : existingListeners) {
-            final ExecutorService executorService = t.v2();
-            final ActionListener<V> listener = t.v1();
-            if (executorService == EsExecutors.DIRECT_EXECUTOR_SERVICE) {
-                notifyListenerDirectly(listener);
-            } else {
-                notifyListener(listener, executorService);
+        final var existingListeners = acquireExistingListeners();
+        if (existingListeners != null) {
+            for (final var listener : existingListeners) {
+                notifyListener(listener);
             }
         }
     }
 
-    private void notifyListenerDirectly(ActionListener<V> listener) {
-        // call get in a non-blocking fashion as we could be on a network thread
-        // or another thread like the scheduler, which we should never block!
-        assert done;
-        ActionListener.completeWith(listener, () -> FutureUtils.get(ListenableFuture.this, 0L, TimeUnit.NANOSECONDS));
+    private synchronized boolean addListenerIfIncomplete(ActionListener<V> listener, Executor executor, ThreadContext threadContext) {
+        // check done under lock since it could have been modified; also protect modifications to the list under lock
+        if (done) {
+            return false;
+        }
+        if (threadContext != null) {
```

Comment:
Minor: these operations could be done outside the lock. Not sure if there's enough contention here to make any difference though

Reply:
I think there isn't all that much contention here in practice, and it's been like this since it was introduced 4 years ago in #37327.